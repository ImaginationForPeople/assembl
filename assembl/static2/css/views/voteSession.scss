.votesession-page {
  h1,
  h2,
  h3 {
    margin-bottom: 0;
  }

  .vote-instructions {
    padding: 0 10px;
    font-size: $font-size-m;
    font-family: $font-family4;
    max-width: 780px;
    line-height: 32px;
  }

  .content-section {
    padding: 0;
  }

  .available-tokens {
    margin: 20px 0 40px;

    .category-available-tokens {
      display: flex;
      padding: 0 0 15px;
      max-width: 780px;
      margin-left: -5px;

      h2 {
        margin-top: 7px;
      }

      .tokens {
        padding-left: 100px;
        padding-top: 9px;
      }

      .text {
        margin: 0;
      }

      span {
        font-size: $font-size-xs;
      }
    }
  }

  .available-tokens-sticky {
    position: fixed;
    top: 60px;
    left: 0;
    right: 0;
    z-index: 100;
    border-top: 1px solid $grey1;
    box-shadow: 3px 4px 7px $grey1;

    .separator {
      display: none;
    }

    .category-available-tokens {
      .tokens {
        padding: 20px 0 0 0;
        display: inline-block;
        vertical-align: top;
        text-align: left;

        svg {
          margin: 0 3px;
        }
      }

      .text {
        line-height: 20px;
        margin: 0 0 5px;
        display: inline-block;
        text-align: left;

        .dark-title-2 {
          font-size: $font-size-s;
          line-height: 18px;
        }

        span {
          font-size: $font-size-xs;
        }
      }
    }
  }

  .extra-margin-top {
    margin-top: 400px;
  }

  .tokens > * {
    margin: 7px 3px;
  }

  .theme-box {
    &:first-child {
      margin: -70px 0 0;
    }
  }

  .proposal.row {
    margin: 90px 10px 70px;

    .proposal-title {
      font-family: $font-family;
      margin-top: 0;
      font-size: 22px;
    }

    .text {
      line-height: 24px;
    }

    .tokens-line {
      margin-bottom: 0;

      p {
        margin: 0;
      }

      .tokens {
        margin-left: -5px;
      }
    }

    .tokens > button {
      border: none;
      padding: 0;
      background: none;

      &:focus {
        outline: none;
      }

      &:active {
        box-shadow: none;
      }
    }

    .tokens > .btn-default:hover {
      background-color: transparent;
    }
  }

  .form-actions {
    margin-top: 50px;
    margin-bottom: 100px;
  }

  .link-button {
    border: none;
    font-size: $font-size-xs;
    color: $link-color;
    text-decoration: none;
    background-color: $grey0;
    padding: 15px;

    &:hover {
      color: $link-hover-color;
      text-decoration: none;
    }

    &:focus {
      outline: none;
    }

    &:active {
      box-shadow: none;
    }

    [class^='assembl-icon-']:before,
    [class*=' assembl-icon-']:before {
      font-size: $font-size-xxs;
      margin-left: 5px;
    }
  }

  .link-button:focus,
  .link-button:active,
  .link-button:hover {
    background-color: transparent;
  }

  .expand-row {
    margin: 0 40px;
  }

  .participants-count {
    .assembl-icon-profil:before {
      font-size: 40px;
      margin-bottom: 8px;
    }
  }

  .participants-count__text {
    color: $first-color;
    padding: 0 10px 20px 10px;
  }

  .proposal-results {
    .participants-count {
      border: 1px solid $grey5;
      margin: 20px 0;
      min-height: 120px;
    }

    .doughnut-container {
      text-align: center;
      height: 150px;
      width: 150px;
      padding: 10px 0 0;
      position: relative;
      margin: auto;

      .doughnut {
        .circle {
          stroke-width: 18px;
        }
      }

      .doughnut-totalCount {
        position: absolute;
        top: 63px;
        left: 53px;
        width: 45px;
        font-size: $font-size-s;

        .vote-totalCount {
          font-size: $font-size-sm;
          font-weight: bold;
          font-family: $font-family3;
        }
      }
    }
  }

  .votes-in-progress {
    margin: 0 40px;
    padding: 0 15px 15px;

    .padding-s {
      padding-left: 5px;
      padding-right: 5px;
    }

    .box {
      border: 1px solid $grey5;
      margin-top: 10px;
      height: 140px;
      text-align: center;
    }

    .tokens-box {
      text-align: center;
    }

    .doughnut-container {
      margin: 0 auto 0 auto;
      height: 80px;
      width: 80px;
      position: relative;

      .doughnut {
        .circle {
          stroke-width: 18px;
        }
      }

      .doughnut-totalCount {
        position: absolute;
        top: 23px;
        left: 24px;
        width: 34px;
        font-size: $font-size-xxs;

        .vote-totalCount {
          font-size: $font-size-xs;
          font-family: $font-family3;
        }
      }
    }
  }

  .expand-result {
    padding-bottom: 20px;
  }

  .gauge-box {
    padding: 20px 0 0;
  }

  .gauge-box__instructions {
    font-size: $font-size-s;
  }

  .gauge-box__estimate {
    color: $first-color;
  }

  .separator {
    border-bottom: 1px $btn-dark-border-style $grey5;
    max-width: 605px;
    margin-left: -5px;
  }
}

.token-tooltip {
  border-radius: 4px;

  .tooltip-inner {
    background: none;
  }
}

@media screen and (min-width: $screen-sm-min) {
  .votesession-page {
    .proposal.row {
      padding: 15px 5px 15px 0;
      display: flex;
    }

    .participants-count {
      height: inherit;
      padding: 20px 0 0;
    }

    .tokens-box {
      padding: 19px 0 19px 0;
      text-align: center;
    }
  }
}

@media screen and (max-width: $screen-xs-max) {
  .votesession-page {
    .votes-in-progress {
      padding: 0 0 15px 0;

      .box {
        margin-top: 15px;
        height: 160px;
      }
    }

    .available-tokens-sticky {
      display: none;
    }

    .container {
      padding: 0;
      margin: 0;
    }

    .proposal.row {
      padding: 15px 0 15px;
    }

    .participants-count {
      padding: 32px 0 0;
    }

    .tokens-box {
      padding: 10px 0 20px 0;
      text-align: center;
    }

    .section-title {
      display: block;
    }

    .max-container {
      padding-left: 15px;
      padding-right: 15px;
    }
  }
}

@media screen and (min-width: $screen-xl-min) {
  .proposal-results {
    width: 80%;
  }
}

@media screen and (max-width: 660px) {
  .votesession-page {
    .available-tokens {
      .category-available-tokens {
        display: block;

        .tokens {
          padding-left: 10px;
          margin-left: -5px;
        }
      }
    }
  }
}

@media screen and (max-width: 420px) {
  .votesession-page {
    .title-section .dark-title-1 {
      margin: 0 5px;
    }
  }
}
