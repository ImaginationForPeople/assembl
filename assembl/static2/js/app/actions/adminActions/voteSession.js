// @flow
import * as actionTypes from '../actionTypes';
import { createRandomId } from '../../utils/globalFunctions';

export const updateVoteSessionPageTitle = (locale: string, value: string): actionTypes.UpdateVoteSessionPageTitle => ({
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_VOTE_SESSION_PAGE_TITLE
});

export const updateVoteSessionPageSeeCurrentVotes = (value: string): actionTypes.UpdateVoteSessionPageSeeCurrentVotes => ({
  value: value,
  type: actionTypes.UPDATE_VOTE_SESSION_PAGE_SEECURRENTVOTES
});

export const updateVoteSessionPageSubtitle = (locale: string, value: string): actionTypes.UpdateVoteSessionPageSubtitle => ({
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_VOTE_SESSION_PAGE_SUBTITLE
});

export const updateVoteSessionPageInstructionsTitle = (
  locale: string,
  value: string
): actionTypes.UpdateVoteSessionPageInstructionsTitle => ({
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_VOTE_SESSION_PAGE_INSTRUCTIONS_TITLE
});

export const updateVoteSessionPageInstructionsContent = (
  locale: string,
  value: string
): actionTypes.UpdateVoteSessionPageInstructionsContent => ({
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_VOTE_SESSION_PAGE_INSTRUCTIONS_CONTENT
});

export const updateVoteSessionPagePropositionsTitle = (
  locale: string,
  value: string
): actionTypes.UpdateVoteSessionPagePropositionsTitle => ({
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_VOTE_SESSION_PAGE_PROPOSITIONS_TITLE
});

export const updateVoteSessionHeaderImage = (value: File): actionTypes.UpdateVoteSessionHeaderImage => ({
  value: value,
  type: actionTypes.UPDATE_VOTE_SESSION_PAGE_IMAGE
});

export const updateVoteModules = (voteModules: actionTypes.VoteModulesArray): actionTypes.UpdateVoteModules => ({
  voteModules: voteModules,
  type: actionTypes.UPDATE_VOTE_MODULES
});

export const createTokenVoteModule = (id: string): actionTypes.CreateTokenVoteModule => ({
  id: id,
  type: actionTypes.CREATE_TOKEN_VOTE_MODULE
});

export const deleteVoteModule = (id: string): actionTypes.DeleteVoteModule => ({
  id: id,
  type: actionTypes.DELETE_VOTE_MODULE
});

export const createGaugeVoteModule = (id: string): actionTypes.CreateGaugeVoteModule => ({
  id: id,
  type: actionTypes.CREATE_GAUGE_VOTE_MODULE
});

export const updateTokenVoteExclusiveCategory = (id: string, value: boolean): actionTypes.UpdateTokenVoteExclusiveCategory => ({
  id: id,
  value: value,
  type: actionTypes.UPDATE_TOKEN_VOTE_EXCLUSIVE_CATEGORY
});

export const updateTokenVoteInstructions = (
  id: string,
  locale: string,
  value: string
): actionTypes.UpdateTokenVoteInstructions => ({
  id: id,
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_TOKEN_VOTE_INSTRUCTIONS
});

export const createTokenVoteCategory = (id: string, moduleId: string): actionTypes.CreateTokenVoteCategory => ({
  id: id,
  moduleId: moduleId,
  type: actionTypes.CREATE_TOKEN_VOTE_CATEGORY
});

export const deleteTokenVoteCategory = (moduleId: string, index: number): actionTypes.DeleteTokenVoteCategory => ({
  moduleId: moduleId,
  index: index,
  type: actionTypes.DELETE_TOKEN_VOTE_CATEGORY
});

export const updateTokenVoteCategoryTitle = (
  id: string,
  locale: string,
  value: string,
  moduleId: string
): actionTypes.UpdateTokenVoteCategoryTitle => ({
  moduleId: moduleId,
  id: id,
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_TOKEN_VOTE_CATEGORY_TITLE
});

export const updateTokenTotalNumber = (id: string, value: number, moduleId: string): actionTypes.UpdateTokenTotalNumber => ({
  moduleId: moduleId,
  id: id,
  value: value,
  type: actionTypes.UPDATE_TOKEN_TOTAL_NUMBER
});

export const updateTokenVoteCategoryColor = (
  id: string,
  value: string,
  moduleId: string
): actionTypes.UpdateTokenVoteCategoryColor => ({
  moduleId: moduleId,
  id: id,
  value: value,
  type: actionTypes.UPDATE_TOKEN_VOTE_CATEGORY_COLOR
});

export const updateGaugeVoteInstructions = (
  id: string,
  locale: string,
  value: string
): actionTypes.UpdateGaugeVoteInstructions => ({
  id: id,
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_GAUGE_VOTE_INSTRUCTIONS
});

export const updateGaugeVoteIsNumber = (id: string, value: boolean): actionTypes.UpdateGaugeVoteIsNumber => ({
  id: id,
  value: value,
  type: actionTypes.UPDATE_GAUGE_VOTE_IS_NUMBER
});

export const updateGaugeVoteNbTicks = (id: string, value: number): actionTypes.UpdateGaugeVoteNbTicks => ({
  id: id,
  value: value,
  type: actionTypes.UPDATE_GAUGE_VOTE_NUMBER_TICKS
});

export const createGaugeVoteChoice = (moduleId: string, id: string): actionTypes.CreateGaugeVoteChoice => ({
  moduleId: moduleId,
  id: id,
  type: actionTypes.CREATE_GAUGE_VOTE_CHOICE
});

export const deleteGaugeVoteChoice = (moduleId: string, index: number): actionTypes.DeleteGaugeVoteChoice => ({
  moduleId: moduleId,
  index: index,
  type: actionTypes.DELETE_GAUGE_VOTE_CHOICE
});

export const updateGaugeVoteChoiceLabel = (
  id: string,
  locale: string,
  value: string,
  moduleId: string
): actionTypes.UpdateGaugeVoteChoiceLabel => ({
  moduleId: moduleId,
  id: id,
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_GAUGE_VOTE_CHOICE_LABEL
});

export const updateGaugeMinimum = (id: string, value: number): actionTypes.UpdateGaugeMinimum => ({
  id: id,
  value: value,
  type: actionTypes.UPDATE_GAUGE_MINIMUM
});

export const updateGaugeMaximum = (id: string, value: number): actionTypes.UpdateGaugeMaximum => ({
  id: id,
  value: value,
  type: actionTypes.UPDATE_GAUGE_MAXIMUM
});

export const updateGaugeUnit = (id: string, value: string): actionTypes.UpdateGaugeUnit => ({
  id: id,
  value: value,
  type: actionTypes.UPDATE_GAUGE_UNIT
});

export const updateVoteSessionPage = (value: Object): actionTypes.UpdateVoteSessionPage => ({
  id: value.id,
  titleEntries: value.titleEntries,
  seeCurrentVotes: value.seeCurrentVotes,
  subTitleEntries: value.subTitleEntries,
  instructionsSectionTitleEntries: value.instructionsSectionTitleEntries,
  instructionsSectionContentEntries: value.instructionsSectionContentEntries,
  propositionsSectionTitleEntries: value.propositionsSectionTitleEntries,
  headerImage: value.headerImage,
  type: actionTypes.UPDATE_VOTE_SESSION_PAGE
});

export const updateVoteProposals = (voteProposals: actionTypes.VoteProposalsArray): actionTypes.UpdateVoteProposals => ({
  voteProposals: voteProposals,
  type: actionTypes.UPDATE_VOTE_PROPOSALS
});

export const createVoteProposal = (id: string): actionTypes.CreateVoteProposal => ({
  id: id,
  type: actionTypes.CREATE_VOTE_PROPOSAL
});

export const deleteVoteProposal = (id: string): actionTypes.DeleteVoteProposal => ({
  id: id,
  type: actionTypes.DELETE_VOTE_PROPOSAL
});

export const updateVoteProposalTitle = (id: string, locale: string, value: string): actionTypes.UpdateVoteProposalTitle => ({
  id: id,
  value: value,
  locale: locale,
  type: actionTypes.UPDATE_VOTE_PROPOSAL_TITLE
});

export const updateVoteProposalDescription = (
  id: string,
  locale: string,
  value: string
): actionTypes.UpdateVoteProposalDescription => ({
  id: id,
  value: value,
  locale: locale,
  type: actionTypes.UPDATE_VOTE_PROPOSAL_DESCRIPTION
});

export const moveProposalUp = (id: string): actionTypes.MoveProposalUp => ({
  id: id,
  type: actionTypes.MOVE_PROPOSAL_UP
});

export const moveProposalDown = (id: string): actionTypes.MoveProposalDown => ({
  id: id,
  type: actionTypes.MOVE_PROPOSAL_DOWN
});

export const addModuleToProposal = (
  id: string,
  proposalId: string,
  voteSpecTemplateId: string
): actionTypes.AddModuleToProposal => ({
  id: id,
  proposalId: proposalId,
  voteSpecTemplateId: voteSpecTemplateId,
  type: actionTypes.ADD_MODULE_TO_PROPOSAL
});

export const createVoteProposalAndModules = (id: string) => (dispatch: Function, getState: Function) => {
  dispatch(createVoteProposal(id));
  const { modulesInOrder } = getState().admin.voteSession;
  modulesInOrder.forEach(moduleId => dispatch(addModuleToProposal(createRandomId(), id, moduleId)));
};

export const undeleteModule = (id: string): actionTypes.UndeleteModule => ({
  id: id,
  type: actionTypes.UNDELETE_MODULE
});

/* if a vote spec template params changes, set _hasChanged to true on all its dependencies */
export const markAllDependenciesAsChanged = (id: string): actionTypes.MarkAllDependenciesAsChanged => ({
  id: id,
  type: actionTypes.MARK_ALL_DEPENDENCIES_AS_CHANGED
});

export const setValidationErrors = (id: string, errors: ValidationErrors): actionTypes.SetValidationErrors => ({
  errors: errors,
  id: id,
  type: actionTypes.SET_VALIDATION_ERRORS
});

export const cancelModuleCustomization = (id: string): actionTypes.CancelModuleCustomization => ({
  id: id,
  type: actionTypes.CANCEL_MODULE_CUSTOMIZATION
});