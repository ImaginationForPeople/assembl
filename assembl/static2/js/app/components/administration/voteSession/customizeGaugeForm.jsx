// @flow
import * as React from 'react';
import { List, Map } from 'immutable';
import { connect } from 'react-redux';
import { Button, Checkbox } from 'react-bootstrap';
import { Translate } from 'react-redux-i18n';

import { DumbGaugeForm, getGaugeModuleInfo, type VoteChoice } from './gaugeForm';
import { createRandomId } from '../../../utils/globalFunctions';
import { cancelAllDependenciesCustomization, updateVoteModule } from '../../../actions/adminActions/voteSession';

type Props = {
  choices: List<VoteChoice>,
  close: Function,
  cancelAllCustomizations: Function,
  updateModule: Function,
  editLocale: string,
  gaugeModuleId: string,
  instructions: string,
  isCustom: boolean,
  isNumberGauge: boolean,
  maximum: ?number,
  minimum: ?number,
  nbTicks: number,
  type: string,
  unit: ?string,
  voteSpecTemplateId: string
};

type State = {
  _hasChanged: boolean,
  applyToAllProposals: boolean,
  gaugeParams: {
    choices: List<Map<string, string>>,
    instructions: string,
    isNumberGauge: boolean,
    maximum: ?number,
    minimum: ?number,
    nbTicks: number,
    unit: ?string
  },
  saving: boolean
};

export class DumbCustomizeGaugeForm extends React.Component<Props, State> {
  static defaultProps = {
    choices: List()
  };

  constructor(props: Props) {
    super(props);
    this.state = {
      _hasChanged: false,
      applyToAllProposals: false,
      gaugeParams: {
        choices: props.choices,
        instructions: props.instructions,
        isNumberGauge: props.isNumberGauge,
        maximum: props.maximum,
        minimum: props.minimum,
        nbTicks: props.nbTicks,
        type: props.type,
        unit: props.unit
      },
      saving: false
    };
  }

  handleSubmit = () => {
    const { cancelAllCustomizations, close, gaugeModuleId, updateModule } = this.props;
    const { gaugeParams } = this.state;
    this.setState({ saving: true });
    if (this.state.applyToAllProposals) {
      const { voteSpecTemplateId } = this.props;
      const params = {
        ...gaugeParams,
        isCustom: false,
        // set random ids to create new choices
        choices: gaugeParams.choices && gaugeParams.choices.map(c => c.set('id', createRandomId()))
      };
      // ensure that the template is updated before to cancel the customizations
      Promise.resolve(updateModule(voteSpecTemplateId, params)).then(() => cancelAllCustomizations(voteSpecTemplateId));
    } else {
      updateModule(gaugeModuleId, { ...gaugeParams, isCustom: true });
    }
    this.setState({ saving: false });
    close();
  };

  toggleApplyToAllProposals = () => {
    this.setState(prevState => ({
      applyToAllProposals: !prevState.applyToAllProposals,
      _hasChanged: true
    }));
  };

  updateInstructions = (value: string): void => {
    this.setState(prevState => ({
      _hasChanged: true,
      gaugeParams: {
        ...prevState.gaugeParams,
        instructions: value
      }
    }));
  };

  createChoice = (newId: string): void => {
    this.setState(prevState => ({
      _hasChanged: true,
      gaugeParams: {
        ...prevState.gaugeParams,
        choices: prevState.gaugeParams.choices.push(Map({ id: newId, title: '' }))
      }
    }));
  };

  deleteChoice = (idx: number): void => {
    this.setState(prevState => ({
      _hasChanged: true,
      gaugeParams: {
        ...prevState.gaugeParams,
        choices: prevState.gaugeParams.choices.delete(idx)
      }
    }));
  };

  updateNbTicks = (value: number): void => {
    this.setState(prevState => ({
      _hasChanged: true,
      gaugeParams: {
        ...prevState.gaugeParams,
        nbTicks: value
      }
    }));
  };

  updateIsNumberGauge = (value: boolean): void =>
    this.setState(prevState => ({
      _hasChanged: true,
      gaugeParams: { ...prevState.gaugeParams, isNumberGauge: value }
    }));

  handleMinChange = (value: number): void =>
    this.setState(prevState => ({
      _hasChanged: true,
      gaugeParams: {
        ...prevState.gaugeParams,
        minimum: value
      }
    }));

  handleMaxChange = (value: number): void =>
    this.setState(prevState => ({
      _hasChanged: true,
      gaugeParams: {
        ...prevState.gaugeParams,
        maximum: value
      }
    }));

  handleUnitChange = (value: string): void =>
    this.setState(prevState => ({
      _hasChanged: true,
      gaugeParams: {
        ...prevState.gaugeParams,
        unit: value
      }
    }));

  handleGaugeChoiceLabelChange = (choiceId: string, value: string): void => {
    this.setState((prevState) => {
      const idxToUpdate = prevState.gaugeParams.choices.findIndex(choice => choice.get('id') === choiceId);

      return {
        _hasChanged: true,
        gaugeParams: {
          ...prevState.gaugeParams,
          choices: prevState.gaugeParams.choices.update(idxToUpdate, choice => choice.set('title', value))
        }
      };
    });
  };

  getNbTicks = (): number => (this.state.gaugeParams.isNumberGauge ? this.state.gaugeParams.nbTicks : this.state.gaugeParams.choices.size);

  render() {
    const { gaugeModuleId } = this.props;
    return (
      <div className="gauge-modal">
        <h3 className="dark-title-3 center">
          <Translate value="administration.gaugeModal.title" />
        </h3>
        <div className="ellipsis-content">
          <Translate value="administration.gaugeModal.subTitle" />
        </div>

        <div className="form">
          <DumbGaugeForm
            id={gaugeModuleId}
            {...this.state.gaugeParams}
            canChangeType={false}
            nbTicks={this.getNbTicks()}
            createChoice={this.createChoice}
            deleteChoice={this.deleteChoice}
            updateNbTicks={this.updateNbTicks}
            handleInstructionsChange={e => this.updateInstructions(e.target.value)}
            handleNumberGaugeCheck={() => this.updateIsNumberGauge(true)}
            handleNumberGaugeUncheck={() => this.updateIsNumberGauge(false)}
            handleMinChange={this.handleMinChange}
            handleMaxChange={this.handleMaxChange}
            handleUnitChange={this.handleUnitChange}
            handleGaugeChoiceLabelChange={this.handleGaugeChoiceLabelChange}
          />
          <Checkbox id="apply-to-all-button" onClick={this.toggleApplyToAllProposals} checked={this.state.applyToAllProposals}>
            <Translate value="administration.gaugeModal.applyToAllProposalsCheckboxLabel" />
          </Checkbox>
          <Button id="modify-gauge-button" className="button-submit button-dark full-size" disabled={!this.state._hasChanged || this.state.saving} onClick={this.handleSubmit}>
            <Translate value="administration.voteProposals.edit" />
          </Button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, { gaugeModuleId, editLocale }) => {
  const { gaugeChoicesById, modulesById } = state.admin.voteSession;
  const pModule = modulesById.get(gaugeModuleId);
  const moduleTemplate = modulesById.get(pModule.get('voteSpecTemplateId'));
  let moduleInfo = getGaugeModuleInfo(moduleTemplate.merge(pModule), gaugeChoicesById, editLocale);
  // replace ids for choices that come from template to be sure to recreate them
  moduleInfo = {
    ...moduleInfo,
    voteSpecTemplateId: pModule.get('voteSpecTemplateId'),
    choices:
      moduleInfo.choices &&
      moduleInfo.choices.map((c) => {
        if (moduleTemplate.has('choices') && moduleTemplate.get('choices').includes(c.get('id'))) {
          return c.set('id', createRandomId());
        }

        return c;
      })
  };

  return {
    ...moduleInfo,
    isCustom: pModule.get('isCustom')
  };
};

const mapDispatchToProps = (dispatch, { editLocale }) => ({
  cancelAllCustomizations: id => dispatch(cancelAllDependenciesCustomization(id)),
  updateModule: (id, params) => dispatch(updateVoteModule(id, editLocale, params))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(DumbCustomizeGaugeForm);