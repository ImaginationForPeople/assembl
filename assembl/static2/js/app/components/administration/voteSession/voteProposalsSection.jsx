// @flow
import * as React from 'react';
import { type List } from 'immutable';
import { connect } from 'react-redux';
import { I18n } from 'react-redux-i18n';
import { OverlayTrigger } from 'react-bootstrap';

import { addVoteProposalTooltip } from '../../common/tooltips';
import { createVoteProposalAndModules } from '../../../actions/adminActions/voteSession';
import SectionTitle from '../sectionTitle';
import VoteProposalForm from './voteProposalForm';
import { createRandomId } from '../../../utils/globalFunctions';

type VoteProposalsSectionProps = {
  addVoteProposal: Function,
  editLocale: string,
  voteProposals: List<string>
};

const DumbVoteProposalsSection = ({ addVoteProposal, editLocale, voteProposals }: VoteProposalsSectionProps) => (
  <div className="vote-proposals-section">
    <div className="admin-box">
      <SectionTitle
        title={I18n.t('administration.voteProposals.sectionTitle')}
        annotation={I18n.t('administration.annotation')}
      />
      <div className="admin-content">
        <form>
          {voteProposals.map((id, index) => (
            <VoteProposalForm key={id} id={id} index={index + 1} editLocale={editLocale} nbProposals={voteProposals.size} />
          ))}
          <OverlayTrigger placement="top" overlay={addVoteProposalTooltip}>
            <div onClick={addVoteProposal} className="plus margin-l">
              +
            </div>
          </OverlayTrigger>
        </form>
      </div>
    </div>
  </div>
);

const mapStateToProps = ({ admin }) => {
  const { voteProposalsById } = admin.voteSession;
  const { editLocale } = admin;
  return {
    voteProposals: voteProposalsById
      .filter(proposal => !proposal.get('_toDelete'))
      .sortBy(proposal => proposal.get('order'))
      .map(proposal => proposal.get('id'))
      .toList(),
    editLocale: editLocale
  };
};

const mapDispatchToProps = dispatch => ({
  addVoteProposal: () => {
    const newProposalId = createRandomId();
    dispatch(createVoteProposalAndModules(newProposalId));
  }
});

export { DumbVoteProposalsSection };

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(DumbVoteProposalsSection);