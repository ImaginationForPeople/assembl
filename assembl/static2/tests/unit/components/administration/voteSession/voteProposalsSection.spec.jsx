import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { List } from 'immutable';
import { voteProposalsInOrder } from './fakeData';

import { DumbVoteProposalsSection } from '../../../../../js/app/components/administration/voteSession/voteProposalsSection';
import '../../../../helpers/setupTranslations';

describe('VoteProposalsSection component', () => {
  const addVoteProposalSpy = jest.fn();
  const refetchVoteSessionSpy = jest.fn();

  it('should render a VoteProposalsSection component without any proposal', () => {
    const props = {
      voteProposals: List(),
      editLocale: 'fr',
      addVoteProposal: addVoteProposalSpy,
      refetchVoteSession: refetchVoteSessionSpy
    };
    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<DumbVoteProposalsSection {...props} />);
    const result = shallowRenderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  it('should render a voteProposalsSection with 2 proposals', () => {
    const props = {
      voteProposals: voteProposalsInOrder,
      editLocale: 'fr',
      addVoteProposal: addVoteProposalSpy,
      refetchVoteSession: refetchVoteSessionSpy
    };
    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<DumbVoteProposalsSection {...props} />);
    const result = shallowRenderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});