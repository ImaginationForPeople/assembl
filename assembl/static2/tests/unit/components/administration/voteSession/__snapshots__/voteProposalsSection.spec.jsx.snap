// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VoteProposalsSection component should render a VoteProposalsSection component without any proposal 1`] = `
<div
  className="vote-proposals-section"
>
  <div
    className="admin-box"
  >
    <SectionTitle
      annotation="* Fields are required."
      title="Configure the proposals associated to the vote modules"
    />
    <div
      className="admin-content"
    >
      <form>
        Immutable.List [
        ]
        <OverlayTrigger
          defaultOverlayShown={false}
          overlay={
            <Tooltip
              bsClass="tooltip"
              id="addProposalTooltip"
              placement="right"
            >
              <Translate
                value="administration.voteProposals.addProposal"
              />
            </Tooltip>
          }
          placement="top"
          trigger={
            Array [
              "hover",
              "focus",
            ]
          }
        >
          <div
            className="plus margin-l"
            onClick={[Function]}
          >
            +
          </div>
        </OverlayTrigger>
      </form>
    </div>
  </div>
</div>
`;

exports[`VoteProposalsSection component should render a voteProposalsSection with 2 proposals 1`] = `
<div
  className="vote-proposals-section"
>
  <div
    className="admin-box"
  >
    <SectionTitle
      annotation="* Fields are required."
      title="Configure the proposals associated to the vote modules"
    />
    <div
      className="admin-content"
    >
      <form>
        Immutable.List [
          <Connect(DumbVoteProposalForm)
            editLocale="fr"
            id={
              Immutable.Map {
                _isNew: false,
                _toDelete: false,
                id: "1234",
                titleEntries: Immutable.List [
                ],
                descriptionEntries: Immutable.List [
                ],
              }
            }
            index={1}
            nbProposals={2}
            refetchVoteSession={[Function]}
          />,
          <Connect(DumbVoteProposalForm)
            editLocale="fr"
            id={
              Immutable.Map {
                _isNew: false,
                _toDelete: false,
                id: "5678",
                titleEntries: Immutable.List [
                ],
                descriptionEntries: Immutable.List [
                ],
              }
            }
            index={2}
            nbProposals={2}
            refetchVoteSession={[Function]}
          />,
        ]
        <OverlayTrigger
          defaultOverlayShown={false}
          overlay={
            <Tooltip
              bsClass="tooltip"
              id="addProposalTooltip"
              placement="right"
            >
              <Translate
                value="administration.voteProposals.addProposal"
              />
            </Tooltip>
          }
          placement="top"
          trigger={
            Array [
              "hover",
              "focus",
            ]
          }
        >
          <div
            className="plus margin-l"
            onClick={[Function]}
          >
            +
          </div>
        </OverlayTrigger>
      </form>
    </div>
  </div>
</div>
`;
